### PR TITLE
Prevent calling decryption in an encryption context and vice versa

### PR DESCRIPTION
**Pull request by levitte**
This recreates #172

### CODE REVIEW

**Comment by levitte** [comment]
I'd appreciate some advice on what branches to cherry pick this to.

**Comment by kroeckx** [comment]
It's one of those changes that might break some applications, but I would argue that they were already broken, so I'm fine with this going to any branch you want.

File: crypto/evp/evp_enc.c [hunk 1]
```diff
@@ -294,8 +294,9 @@ int is_partially_overlapping(const void *ptr1, const void *ptr2, int len)
     return overlapped;
 }
 
-int EVP_EncryptUpdate(EVP_CIPHER_CTX *ctx, unsigned char *out, int *outl,
-                      const unsigned char *in, int inl)
+static int evp_EncryptDecryptUpdate(EVP_CIPHER_CTX *ctx,
+                                    unsigned char *out, int *outl,
+                                    const unsigned char *in, int inl)
```
[the remark this answers is not in the document]
Comment (levitte) [reply crypto/evp/evp_enc.c]: Ah, I forgot to fix up EVP_DecryptUpdate

**Review by kroeckx — APPROVED** [review]

File: crypto/evp/evp_enc.c [hunk 2]
```diff
@@ -371,6 +372,19 @@ int EVP_EncryptUpdate(EVP_CIPHER_CTX *ctx, unsigned char *out, int *outl,
     return 1;
 }
 
+
+int EVP_EncryptUpdate(EVP_CIPHER_CTX *ctx, unsigned char *out, int *outl,
+                      const unsigned char *in, int inl)
+{
+    /* Prevent accidental use of decryption context when encryption */
```
Comment (mspncp) [inline crypto/evp/evp_enc.c]: grammar: "when encryption"

should be either: "during encryption" or "while encrypting"

(multiple locations)

**Comment by owlstead** [comment]
Thank you for picking this up again! I was kind of hoping not having to recreate an up-to-date linux machine for developing.

File: crypto/evp/evp_enc.c (hunk 1 above)
Comment (mspncp) [inline crypto/evp/evp_enc.c]: Did not check for precedents, but wouldn't an all-lowercase name be more appropriate for an internal function? Something like `[evp_]encrypt_decrypt_update()` ?

File: crypto/evp/evp_enc.c (hunk 2 above)
Comment (levitte) [reply crypto/evp/evp_enc.c]: Fixed

File: crypto/evp/evp_enc.c (hunk 1 above)
Comment (levitte) [reply crypto/evp/evp_enc.c]: I dunno, seriously...  lower case prefix, but then follow the style of the file, was my line of thinking.

**Review by paulidale — APPROVED** [review]
LGTM

File: crypto/evp/evp_enc.c (hunk 1 above)
Comment (mspncp) [reply crypto/evp/evp_enc.c]: Probably you're right. The style of the file (mixed_underscore_withCamelCase) is the problem ;-)

**Comment by levitte** [comment]
I plan on cherry-picking this to 1.1.1 and 1.0.2.  Yay?  Nay?

**Comment by mspncp** [comment]
Yay!

**Comment by levitte** [comment]
Merged.

master:
f2f734d4f9e34643a1d3e5b79d2447cd643519f8 make update
a8bf2f8f24bda57c860e4356f31264fff63fe8b9 Prevent calling decryption in an encryption context and vice versa

**Comment by levitte** [comment]
1.1.1:
4bd0db1feaaf97fbc2bd31f54f1fbdeab80b2b1a make update
fa75ee1aecda0da96216440aa4ea91d4a17e8244 Prevent calling decryption in an encryption context and vice versa

**Comment by levitte** [comment]
1.1.1:
4bd0db1feaaf97fbc2bd31f54f1fbdeab80b2b1a make update
fa75ee1aecda0da96216440aa4ea91d4a17e8244 Prevent calling decryption in an encryption context and vice versa

**Comment by levitte** [comment]
Argh, will have to make a separate PR for 1.0.2...  doesn't cherry-pick cleanly.

**Comment by levitte** [comment]
Backport to 1.0.2 in #7856

**Pull request closed without merging**